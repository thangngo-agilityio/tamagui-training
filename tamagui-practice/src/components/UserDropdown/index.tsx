import { Menu, MenuButton, MenuItem, MenuList } from '@chakra-ui/react';

// Components
import { Avatar } from '../common';

// Icons
import { Logout } from '@/icons';

type TUserDropdown = {
  src?: string;
  alt?: string;
  onClick: () => void;
};

const UserDropdown = ({ src, alt, onClick }: TUserDropdown) => (
  <Menu>
    <MenuButton as="button" aria-label="Options">
      <Avatar src={src} alt={alt} />
    </MenuButton>
    <MenuList>
      <MenuItem icon={<Logout />} onClick={onClick}>
        Logout
      </MenuItem>
    </MenuList>
  </Menu>
);

export default UserDropdown;
