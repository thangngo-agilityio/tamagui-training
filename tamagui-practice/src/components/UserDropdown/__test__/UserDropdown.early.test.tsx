import { fireEvent, render, screen } from '@testing-library/react';
import React from 'react';
import '@testing-library/jest-dom';

// Components
import UserDropdown from '../index';

// Mocking the dependencies
jest.mock('@chakra-ui/react', () => ({
  Menu: ({ children }: { children: React.ReactNode }) => <div>{children}</div>,
  MenuButton: ({ children }: { children: React.ReactNode }) => (
    <button>{children}</button>
  ),
  MenuItem: ({
    children,
    onClick,
  }: {
    children: React.ReactNode;
    onClick: () => void;
  }) => <div onClick={onClick}>{children}</div>,
  MenuList: ({ children }: { children: React.ReactNode }) => (
    <div>{children}</div>
  ),
}));

jest.mock('../../common', () => ({
  Avatar: ({ src, alt }: { src?: string; alt?: string }) => (
    <img src={src} alt={alt} />
  ),
}));

jest.mock('@/icons', () => ({
  Logout: () => <svg>Logout Icon</svg>,
}));

describe('UserDropdown() UserDropdown method', () => {
  // Arrange
  const src = 'avatar.png';
  const alt = 'User Avatar';
  const onClick = jest.fn();
  it('should match snapshot for UserDropdown', () => {
    const { container } = render(
      <UserDropdown src={src} alt={alt} onClick={onClick} />,
    );

    expect(container).toMatchSnapshot();
  });

  it('should render the Avatar with the correct src and alt attributes', () => {
    // Act
    render(<UserDropdown src={src} alt={alt} onClick={onClick} />);

    // Assert
    const avatar = screen.getByAltText(alt);
    expect(avatar).toBeInTheDocument();
    expect(avatar).toHaveAttribute('src', src);
  });

  it('should call onClick when the Logout menu item is clicked', () => {
    // Arrange
    const onClick = jest.fn();

    // Act
    render(<UserDropdown onClick={onClick} />);
    const logoutItem = screen.getByText('Logout');
    fireEvent.click(logoutItem);

    // Assert
    expect(onClick).toHaveBeenCalledTimes(1);
  });
});
