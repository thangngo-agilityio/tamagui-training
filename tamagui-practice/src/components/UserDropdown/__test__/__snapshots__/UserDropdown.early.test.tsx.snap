// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserDropdown() UserDropdown method should match snapshot for UserDropdown 1`] = `
<div>
  <div>
    <button>
      <img
        alt="User Avatar"
        src="avatar.png"
      />
    </button>
    <div>
      <div>
        Logout
      </div>
    </div>
  </div>
</div>
`;
